### PR TITLE
[DPE-3721] - chore: add cruise-control to snap

### PR DESCRIPTION
**Pull request by marcoppenheimer**
##### Review Notes
- Charmed Kafka needs changes to support internal clients for the Metrics Reporter, so this won't work until they're added. Have tested with a locally patched charm.
- This doesn't have tests, and won't run without 'additional' config. We will need to add tests + default config that works with them so that they snap app can work stand-alone.

## MVP CruiseControl addition to Snap
The final layout will be similar to Kafka, as follows:
- `$SNAP/opt/cruise-control/bin` - Modified start script from upstream, and Snap wrapper for setting default env-vars
- `$SNAP/opt/cruise-control/libs` - All `build/libs` and `build/dependant-libs` for running **cruisecontrol**. Also includes libs for **cruise-control-metrics-reporter**, not sure if these are needed or not yet
- `$SNAP_DATA/etc/cruise-control` - Config files. May include:

    - `log4j.properties` - Bundled with snap
    - `cruisecontrol.properties` - Added by charm
    - `cruisecontrol.jaas` - Added by charm
    - `capacity.json` - Added by charm

- `$SNAP_COMMON/var/log/cruise-control` - Dir for logs from Log4j2
- `$SNAP_COMMON/var/lib/cruise-control` - Data dir, but will likely not be used

#### Minor changes to existing Kafka snap
We `unset KAFKA_LOG4J_OPTS` now in `kafka/bin/bin-wrapper.bash`. This avoids very distracting and mostly unnecessary logging when running bin-scripts locally.

We also now take the **STAGE**'d CruiseControl libs and add the metrics-reporter JARs to Kafka's `libs`.

#### Heavily modified `kafka-cruise-control-start.sh`
CruiseControl's default ['start-the-daemon' bash script](https://github.com/linkedin/cruise-control/blob/migrate_to_kafka_2_5/kafka-cruise-control-start.sh) expects that it's running directly from within the CruiseControl repo after being built from source. As a result, it hard-codes a bunch of paths that we have modified to get it to fit in the Snap layout.

tl;dr - We change `base_dir` and set `$CLASSPATH` to include files in a single `/opt/cruise-control/libs` dir. If we compare the diff to upstream:
```diff
30c30
< base_dir=$(dirname $0)
---
> base_dir=$(dirname $0)/..
40,63d39
< # run ./gradlew copyDependantLibs to get all dependant jars in a local dir
< shopt -s nullglob
< for dir in "$base_dir"/cruise-control/build/dependant-libs;
< do
<   if [ -z "$CLASSPATH" ] ; then
<     CLASSPATH="$dir/*"
<   else
<     CLASSPATH="$CLASSPATH:$dir/*"
<   fi
< done
< 
< if [ -z "$CLASSPATH" ]; then
<   CLASSPATH="$base_dir/cruise-control/build/libs/*"
< else
<   CLASSPATH="$CLASSPATH:$base_dir/cruise-control/build/libs/*"
< fi
< 
< if [ -z "$CLASSPATH" ]; then
<   CLASSPATH="$base_dir/cruise-control-metrics-reporter/build/libs/*"
< else
<   CLASSPATH="$CLASSPATH:$base_dir/cruise-control-metrics-reporter/build/libs/*"
< fi
< shopt -u nullglob
< 
143a120,125
> 
> # classpath addition for release
> for file in "$base_dir"/libs/*;
> do
>     CLASSPATH="$CLASSPATH":"$file"
> done
```
#### SASL Auth with JAAS
In order to pick up auth for Kafka + ZooKeeper, similar to Kafka we need to set `-Djava.security.auth.login.config` in `$KAFKA_OPTS`, pointing to a path for some `cruisecontrol.jaas`. For Charmed Kafka, this will be in `/etc/environment` most likely.

An example file would look like:
```jaas
# ZooKeeper auth
Client {
    org.apache.zookeeper.server.auth.DigestLoginModule required
    username="relation-6"
    password="0NMZsBGhd6yoQLVhUNu03LliVdB4HErW";
};

# Kafka auth
KafkaClient {
    org.apache.kafka.common.security.scram.ScramLoginModule required
    username="admin"
    password="GwNjX8CTd0VHXBEoVlUz1ObWB3LaAOi8";
};
````
#### Hard-coded logging path override
We provide a default Log4j2 config packaged with the snap that takes `-Dcruisecontrol.logs.dir` as a Java config, as the default config file provided by upstream is hard-coded to a path we don't want to use.

```diff
8c4
< property.filename=./logs
---
> property.filename=${sys:cruisecontrol.logs.dir}
```

> **_NOTE:_**  CruiseControl uses Log4J2, which has a slightly different syntax to what is already present in the snap for Kafka itself

This gets set in `$KAFKA_LOG4J_OPTS` in the `start-wrapper.bash`, using the snap daemon environment variable `$LOG_DIR`.

We also pass `-Dlog4j.configurationFile` here too, pointing to the file in `$SNAP_DATA/etc/cruise-control` 

#### Minimal properties that need to be set over defaults
```diff
11c11,13
< bootstrap.servers=localhost:9092
---
> bootstrap.servers=10.137.147.101:9092,10.137.147.207:9092,10.137.147.208:9092
> sasl.mechanism=SCRAM-SHA-512
> security.protocol=SASL_PLAINTEXT
90c92
< capacity.config.file=config/capacityJBOD.json
---
> capacity.config.file=/var/snap/charmed-kafka/current/etc/cruise-control/capacity.json
177c179
< zookeeper.connect=localhost:2181/
---
> zookeeper.connect=10.137.147.206:2181,10.137.147.231:2181,10.137.147.46:2181/kafka
180c182
< zookeeper.security.enabled=false
---
> zookeeper.security.enabled=true 
```

#### Proof it works (probably)
```
~  ❯ curl http://localhost:9090/kafkacruisecontrol/state
MonitorState: {state: RUNNING(9.600% trained), NumValidWindows: (4/5) (80.000%), NumValidPartitions: 117/117 (100.000%), flawedPartitions: 0}
ExecutorState: {state: NO_TASK_IN_PROGRESS}
AnalyzerState: {isProposalReady: true, readyGoals: [NetworkInboundUsageDistributionGoal, CpuUsageDistributionGoal, PotentialNwOutGoal, LeaderReplicaDistributionGoal, NetworkInboundCapacityGoal, LeaderBytesInDistributionGoal, DiskCapacityGoal, ReplicaDistributionGoal, RackAwareGoal, TopicReplicaDistributionGoal, NetworkOutboundCapacityGoal, CpuCapacityGoal, DiskUsageDistributionGoal, NetworkOutboundUsageDistributionGoal, ReplicaCapacityGoal]}
AnomalyDetectorState: {selfHealingEnabled:[], selfHealingDisabled:[DISK_FAILURE, BROKER_FAILURE, GOAL_VIOLATION, METRIC_ANOMALY, TOPIC_ANOMALY, MAINTENANCE_EVENT], selfHealingEnabledRatio:{DISK_FAILURE=0.0, BROKER_FAILURE=0.0, GOAL_VIOLATION=0.0, METRIC_ANOMALY=0.0, TOPIC_ANOMALY=0.0, MAINTENANCE_EVENT=0.0}, recentGoalViolations:[], recentBrokerFailures:[], recentMetricAnomalies:[], recentDiskFailures:[], recentTopicAnomalies:[], recentMaintenanceEvents:[], metrics:{meanTimeBetweenAnomalies:{GOAL_VIOLATION:0.00 milliseconds, BROKER_FAILURE:0.00 milliseconds, METRIC_ANOMALY:0.00 milliseconds, DISK_FAILURE:0.00 milliseconds, TOPIC_ANOMALY:0.00 milliseconds}, meanTimeToStartFix:0.00 milliseconds, numSelfHealingStarted:0, numSelfHealingFailedToStart:0, ongoingAnomalyDuration=0.00 milliseconds}, ongoingSelfHealingAnomaly:None, balancednessScore:100.000}
```

### CODE REVIEW

**Review by Batalex — COMMENTED** [review]
I like how clean the diff is, especially in the installation hook and the snapcraft file.
Do we need to add CC's license in the directory with the same name? 

Out of scope question, should we remove kafka-export license since it is not part of the snap?

File: snap/local/opt/cruise-control/bin/kafka-cruise-control-start.sh
Comment (deusebio) [inline snap/local/opt/cruise-control/bin/kafka-cruise-control-start.sh:1]: **question** Did we write this file or did we fetch this from upstream? This seems very low level. Maybe it should go with the source or pulled during the build (in order to limit maintainability and divergence)? I would in general keep files here rather simple (like wrappers) or custom defaults (if we need special configuration, e.g. setting log level)

I have not reviewed this assuming it is from upstream

File: snap/local/opt/kafka/bin/bin-wrapper.bash [hunk 1]
```diff
@@ -3,9 +3,7 @@
 set -e
 
 unset KAFKA_JMX_OPTS
+unset KAFKA_LOG4J_OPTS
 
-if [ "x$KAFKA_LOG4J_OPTS" = "x" ]; then
-    export KAFKA_LOG4J_OPTS="-Dlog4j.configuration=file:${SNAP_DATA}/etc/kafka/tools-log4j.properties"
-fi
```
Comment (deusebio) [inline snap/local/opt/kafka/bin/bin-wrapper.bash:9]: **todo** why is this needed? I don't think we should do this. We need this to make sure that CLI commands are logged out at console, such that when charm does something we get relevant log in the charm logs

File: snap/local/etc/cruise-control/log4j.properties [hunk 2]
```diff
@@ -0,0 +1,60 @@
+rootLogger.level=INFO
+appenders=console, kafkaCruiseControlAppender, operationAppender, requestAppender
+
+property.filename=${sys:cruisecontrol.logs.dir}
```
Comment (deusebio) [inline snap/local/etc/cruise-control/log4j.properties]: **niptick** This does not seem a system property as [these](https://docs.oracle.com/javase/tutorial/essential/environment/sysprop.html). I checked [here](https://logging.apache.org/log4j/3.x/manual/configuration.html#PropertySubstitution) and "main" seems possibly more appropriate. 

Kafka does not use any, so maybe we can use the same. It would definitely be good to have a consistent UX here across. On this, we could actually allow log level tuning using the same property `charmed.kafka.log.level`

**Review by deusebio — COMMENTED** [review]
Changes look sane. I only have a couple of points I'd like to follow before approving

File: snap/local/etc/cruise-control/log4j.properties (hunk 2 above)
Comment (marcoppenheimer) [reply snap/local/etc/cruise-control/log4j.properties]: Ah, I didn't realise there were more labels than just `sys`. I'll find a better option, probably `main`.

I have a suspicion that Kafka does something different with the configuration file for backwards compatibility, as it didn't work without this syntax. I can see if Kafka works with it also.

And yes, agreed on log-level, but again it wasn't working so I've left it out, but it's on the todo list for later!

File: snap/local/opt/kafka/bin/bin-wrapper.bash (hunk 1 above)
Comment (marcoppenheimer) [reply snap/local/opt/kafka/bin/bin-wrapper.bash:9]: I don't agree. Right now, we get hundreds of lines of Java logging, for a single output line. It's been really painful for our users who have manually `unset` the env-var for their deployments.

We don't need the full Java logs for bin-commands, just sysout.

File: snap/local/opt/cruise-control/bin/kafka-cruise-control-start.sh
Comment (deusebio) [reply snap/local/opt/cruise-control/bin/kafka-cruise-control-start.sh:1]: Sorry, my bad, just read the PR description where this was commented thoroughly (and thanks for the very detailed description! **praise**). We should probably built this from source and this patch would nicely go in the first commit enabling build from source. It is fine for now, but eventually we could keep this there.

Comment (marcoppenheimer) [reply snap/local/opt/cruise-control/bin/kafka-cruise-control-start.sh:1]: It's in the PR description, but it's from the source, but patched.

File: snap/local/opt/kafka/bin/bin-wrapper.bash (hunk 1 above)
Comment (deusebio) [reply snap/local/opt/kafka/bin/bin-wrapper.bash:9]: > This avoids very distracting and mostly unnecessary logging when running bin-scripts locally.

I don't think it is distracting and in general, I would REALLY prefer to keep it as-is. I prefer being safer and have more granular logs (in the appropriate place) rather than less. Also, the usage of that logging file (with that name) is consistent with upstream, so I don't think this is very weird.

File: snap/local/etc/cruise-control/log4j.properties (hunk 2 above)
Comment (deusebio) [reply snap/local/etc/cruise-control/log4j.properties]: Oh, that's interesting if it did not work. Are they (kafka and CC) using the same logging library? I suppose the variable resolution is handled there. If this syntax works on Kafka, let's use this one instead (possibly with main) and add the log level here as well, just to keep consistency

**Comment by marcoppenheimer** [comment]
Since last review:

#### `chore: add licenses`
#### `chore: add default cruisecontrol.properties file`
#### `chore: update log4j2 configuration`
- `sys` seems to be necessary
- Removed console appender from writing to syslogs

TODO:
- Open Kafka PR passing `KAFKA_LOG4J_OPTS` to `run_bin_command`

**Comment by marcoppenheimer** [comment]
Have re-requested review as it was a bit stale.
Since last review:

#### `chore: add licenses`
#### `chore: add default cruisecontrol.properties file`
#### `chore: update log4j2 configuration`
- `sys` seems to be necessary
- Removed console appender from writing to syslogs

Have also opened a PR on the Kafka repo using the `KAFKA_LOG4J_OPTS` where necessary - https://github.com/canonical/kafka-operator/pull/201

File: snap/local/etc/cruise-control/log4j.properties [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+appender.requestAppender.filePattern=${filename}/kafkacruisecontrol-request.log.%d{yyyy-MM-dd-HH}
+appender.requestAppender.layout.type=PatternLayout
+appender.requestAppender.layout.pattern=[%d] %p %m (%c)%n
+appender.requestAppender.policies.type=Policies
+appender.requestAppender.policies.time.type=TimeBasedTriggeringPolicy
+appender.requestAppender.policies.time.interval=1
+
+# Loggers
+logger.cruisecontrol.name=com.linkedin.kafka.cruisecontrol
+logger.cruisecontrol.level=${main:cruisecontrol.log.level}
```
Comment (deusebio) [inline snap/local/etc/cruise-control/log4j.properties]: **todo** update with `sys` if `main` was not working

File: snap/local/etc/cruise-control/log4j.properties [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+logger.cruisecontrol.appenderRef.kafkaCruiseControlAppender.ref=kafkaCruiseControlFile
+
+logger.detector.name=com.linkedin.kafka.cruisecontrol.detector
+logger.detector.level=${main:cruisecontrol.log.level}
```
Comment (deusebio) [inline snap/local/etc/cruise-control/log4j.properties]: **todo** same here

File: snap/local/etc/cruise-control/log4j.properties [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+logger.detector.appenderRef.kafkaCruiseControlAppender.ref=kafkaCruiseControlFile
+
+logger.operationLogger.name=operationLogger
+logger.operationLogger.level=${main:cruisecontrol.log.level}
```
Comment (deusebio) [inline snap/local/etc/cruise-control/log4j.properties]: **todo** same here

File: snap/local/etc/cruise-control/log4j.properties [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+logger.operationLogger.appenderRef.operationAppender.ref=operationFile
+
+logger.CruiseControlPublicAccessLogger.name=CruiseControlPublicAccessLogger
+logger.CruiseControlPublicAccessLogger.level=${main:cruisecontrol.log.level}
```
Comment (deusebio) [inline snap/local/etc/cruise-control/log4j.properties]: **todo** same here

File: snap/local/opt/kafka/bin/bin-wrapper.bash (hunk 1 above)
Comment (deusebio) [reply snap/local/opt/kafka/bin/bin-wrapper.bash:9]: **issue** If we get rid of these lines, which log4j configuration would we use for any commands we run with the snap? The reason for these lines to be there is that when running the Kafka bin commands with the snap one would use the `tools-log4j.properties` as for upstream. However, differently from upstream, this file is in another directory now (since the default one would be in the read-only squashfs)

It is also what we do in the other scripts

File: snap/local/etc/cruise-control/log4j.properties [hunk 7]
```diff
@@ -0,0 +1,60 @@
+#
+# Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+#
+
+cruisecontrol.log.level=INFO
+
+rootLogger.level=${sys:cruisecontrol.log.level}
+appenders=kafkaCruiseControlAppender, operationAppender, requestAppender
+
+property.filename=${sys:cruisecontrol.logs.dir}
```
Comment (deusebio) [inline snap/local/etc/cruise-control/log4j.properties:10]: **niptick** maybe **logdir**? `filename` seems a bit misleading here

**Review by deusebio — COMMENTED** [review]
I have still few comments I'd like to get addressed. Also, thanks for investigating the `main` for variable handling, I'm a bit disappointed that it does not work :(, but that's all right! Thanks for investingating this! :man_shrugging: 

But before merging, I also believe it would be very important to

1. Starting cruise control in CI and doing the curl (that you mention in the PR description) to make sure the service is up and running. The defaults in the CC properties file - e.g. zookeeper and kafka endpoint - should be such that the plain "deployment" works out-of-the box (like what we have in CI for Kafka and ZooKeeper)

2. Just add the same steps in the `README.md`, to keep the documentation up to date.

**Review by Batalex — APPROVED** [review]
LGTM

File: snap/local/etc/cruise-control/log4j.properties (hunk 7 above)
Comment (marcoppenheimer) [reply snap/local/etc/cruise-control/log4j.properties:10]: I didn't choose it, went with whatever was the default in the config.

File: snap/local/etc/cruise-control/log4j.properties (hunk 3 above)
Comment (deusebio) [reply snap/local/etc/cruise-control/log4j.properties]: why was this resolved? can you clarify where/why sometimes we use `sys`, and sometime we use `main`?

File: snap/local/etc/cruise-control/log4j.properties [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+appender.kafkaCruiseControlAppender.type=RollingFile
+appender.kafkaCruiseControlAppender.name=kafkaCruiseControlFile
+appender.kafkaCruiseControlAppender.fileName=${filename}/kafkacruisecontrol.log
+appender.kafkaCruiseControlAppender.filePattern=${filename}/kafkacruisecontrol.log.%d{yyyy-MM-dd-HH}
+appender.kafkaCruiseControlAppender.layout.type=PatternLayout
+appender.kafkaCruiseControlAppender.layout.pattern=[%d] %p %m (%c)%n
+appender.kafkaCruiseControlAppender.policies.type=Policies
+appender.kafkaCruiseControlAppender.policies.time.type=TimeBasedTriggeringPolicy
+appender.kafkaCruiseControlAppender.policies.time.interval=1
```
Comment (zmraul) [inline snap/local/etc/cruise-control/log4j.properties]: I suggest to keep the same approach as we recently did with Kafka:
```suggestion
appender.kafkaCruiseControlAppender.policies.size.type = SizeBasedTriggeringPolicy
appender.kafkaCruiseControlAppender.policies.size.size=100MB
appender.kafkaCruiseControlAppender.strategy.type = DefaultRolloverStrategy
appender.kafkaCruiseControlAppender.strategy.max = 10
```

File: snap/local/etc/cruise-control/log4j.properties [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+appender.operationAppender.type=RollingFile
+appender.operationAppender.name=operationFile
+appender.operationAppender.fileName=${filename}/kafkacruisecontrol-operation.log
+appender.operationAppender.filePattern=${filename}/kafkacruisecontrol-operation.log.%d{yyyy-MM-dd}
+appender.operationAppender.layout.type=PatternLayout
+appender.operationAppender.layout.pattern=[%d] %p [%c] %m %n
+appender.operationAppender.policies.type=Policies
+appender.operationAppender.policies.time.type=TimeBasedTriggeringPolicy
+appender.operationAppender.policies.time.interval=1
```
Comment (zmraul) [inline snap/local/etc/cruise-control/log4j.properties]:
```suggestion
appender.operationAppender.policies.size.type = SizeBasedTriggeringPolicy
appender.operationAppender.policies.size.size=100MB
appender.operationAppender.strategy.type = DefaultRolloverStrategy
appender.operationAppender.strategy.max = 10
```

File: snap/local/etc/cruise-control/log4j.properties [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+appender.requestAppender.type=RollingFile
+appender.requestAppender.name=requestFile
+appender.requestAppender.fileName=${filename}/kafkacruisecontrol-request.log
+appender.requestAppender.filePattern=${filename}/kafkacruisecontrol-request.log.%d{yyyy-MM-dd-HH}
+appender.requestAppender.layout.type=PatternLayout
+appender.requestAppender.layout.pattern=[%d] %p %m (%c)%n
+appender.requestAppender.policies.type=Policies
+appender.requestAppender.policies.time.type=TimeBasedTriggeringPolicy
+appender.requestAppender.policies.time.interval=1
```
Comment (zmraul) [inline snap/local/etc/cruise-control/log4j.properties]:
```suggestion
appender.requestAppender.policies.size.type = SizeBasedTriggeringPolicy
appender.requestAppender.policies.size.size=100MB
appender.requestAppender.strategy.type = DefaultRolloverStrategy
appender.requestAppender.strategy.max = 10
```

File: snap/local/opt/kafka/bin/bin-wrapper.bash (hunk 1 above)
Comment (zmraul) [reply snap/local/opt/kafka/bin/bin-wrapper.bash:9]: I agree with Marc here. For bin commands it's quite ugly to show a diarrhea of Java code which adds +100 lines at times when the output should be a simple line. That's why `--verbose` flags exist on CLIs.

Maybe we can implement a `--verbose` flag, and check for that on this script?

File: snap/snapcraft.yaml [hunk 11]
```diff
@@ -19,25 +19,41 @@ hooks:
...
       read:
         - $SNAP_COMMON/var/log/kafka
+  cc-logs:
+    interface: content
+    source:
+      read:
+        - $SNAP_COMMON/var/log/cruise-control
 
 apps:
   daemon:
```
Comment (zmraul) [inline snap/snapcraft.yaml:36]: nit: I would change this to be `kafka`. Makes things explicit once we start having multiple apps on the snap.

**Review by zmraul — APPROVED** [review]
Very nice! I hope the script to start cc wasn't a pain to get working :grimacing:

File: snap/local/etc/cruise-control/log4j.properties (hunk 8 above)
Comment (zmraul) [reply snap/local/etc/cruise-control/log4j.properties]: After checking, there are up to 6 files between kafka and cc. A fully running cluster log output might be too much if each unit is going to have up to 3Gb (uncompressed) of logs.

File: snap/snapcraft.yaml (hunk 11 above)
Comment (marcoppenheimer) [reply snap/snapcraft.yaml:36]: Let's leave it for now and update it to something useful when we have KRaft and need to also update it everywhere in the charms and docs, but I see your point.

**Review by deusebio — APPROVED** [review]
LGTM! Thanks!
